File: Biot.jl-main/src/FieldCalc.jl
## Biot savart law
# using LinearAlgebra
# using PyPlot
# using MPI_Tools
# using Interpolations


"""
Makes a point path of an ellipse with r₁,r₂ as two radii
normal direction is assumed to be in Z
kwargs: Center is the center of path [X,Y,Z]
NPts is the number of points along path


"""
function MakeEllip(r₁,r₂;Center = [0,0,0],NPts=100)
    X,Y,Z = ([r₁ .* cos.(I/NPts *2*π) for I in 1:NPts].+Center[1],
             [r₂ .* sin.(I/NPts *2*π) for I in 1:NPts].+Center[2],
             [0 for I in 1:NPts].+Center[3])
    
    Coords = hcat(X,Y,Z)

end


"""
Simple analytical formula for the field on axis of a Circle
    R - radius in meters
    z - axial offset in meters
    
    kwargs:
        I = current in coil

"""
function FieldOnAxis_Circ(R,z;I=1)
    μ₀ = 4*pi*1e-7
    Bz = μ₀*2*pi*R^2*I ./ (4*π* (z^2 + R^2)^(3/2))
end

# x = -.3:.01:.3
# y = x
# DefField = [BiotSav(MakeEllip(1,1;Center = [0,0,0],NPts=1000),[i,j,k];MinThreshold= 0.05)[3]
#             for i in x,
#                 j in y,
#                 k=0]
# X,Y = MPI_Tools.meshgrid(x,y)
# PyPlot.surf(X[:],Y[:],DefField[:],cmap="jet")

function MakeEllipTestPoints(r₁,r₂;Center = [0,0,0],NPts=6,Layers = 3)
    LL=1
    X,Y,Z = ([r₁*LL/(Layers+1) .* cos.(I/NPts *2*π) for I in 1:NPts].+Center[1],
             [r₂ *LL/(Layers+1) .* sin.(I/NPts *2*π) for I in 1:NPts].+Center[2],
             [0 for I in 1:NPts].+Center[3])
             Coords = hcat(X,Y,Z)
    for LL in 2:Layers
        X₁,Y₁,Z₁ = ([r₁*LL/(Layers+1) .* cos.(I/NPts *2*π) for I in 1:NPts].+Center[1],
                 [r₂ *LL/(Layers+1) .* sin.(I/NPts *2*π) for I in 1:NPts].+Center[2],
                 [0 for I in 1:NPts].+Center[3])
                 Coords = hcat(X,Y,Z)
        X = vcat(X,X₁)
        Y = vcat(Y,Y₁)
        Z = vcat(Z,Z₁)

     end
     return X,Y,Z
 end


"""
This function takes in a pointpath [X,Y,Z] array and maps the interior

Inputs are
PointPath [X,Y,Z] , N x 3 array
NumLayers - number of interior test points

"""
 function FieldMapPointPath(PointPath, NumLayers; WeightRadius = false,InvWeights = false, PlotAxes = nothing)

     xTP = PointPath[:,1]
     x   = PointPath[:,1]
     yTP = PointPath[:,2]
     y   = PointPath[:,2]
     x₁ = minimum(xTP)
     x₂ = maximum(xTP)
     y₁ = minimum(yTP)
     y₂ = maximum(yTP)
     xmid = (x₁+ x₂)/2
     ymid = (y₁+ y₂)/2



     for i in 1:(NumLayers-2)
         xTP = vcat(xTP,(x .- xmid) .* (i/NumLayers) .+ xmid)
         yTP = vcat(yTP,(y .- ymid) .* (i/NumLayers) .+ ymid)
     end
     TPList = hcat(xTP,yTP,zeros(size(yTP)))
     TP_Arr =[([TPList[i,1],TPList[i,2],TPList[i,3]]) for i in 1:length(xTP)]
     if WeightRadius
         Weights = PointPath[:,1]
     else
          Weights = ones(size( PointPath[:,1]))
      end
      if InvWeights
          Weights = 1 ./Weights
      end
      push!(Weights,Weights[1])
     BMag = [BiotSav(PointPath,TP_Arr[i];Current = Weights,MinThreshold =1e-8)[3] for i in 1:length(TP_Arr)]

     BMag = abs.(BMag)

     BPlot = findall(xx-> xx!=0,BMag)
     if PlotAxes!== nothing
        PyPlot.axes(PlotAxes)
     end

     surf(xTP[BPlot],yTP[BPlot],BMag[BPlot],cmap="jet")


 end



function WireNearField_B(r,Wire_R;I=1)
    μ₀ = 4*π*1e-7
    if r<Wire_R
        B = μ₀*r*I / (2*π*Wire_R^2)
    elseif r==Wire_R
        B = μ₀*I / (2*π*Wire_R)
    else
        B = μ₀*I / (2*π*r)
    end
    return B
end

function DerivField(r,Wire_R=0.001;I=1)
    μ₀ = 4*π*1e-7
    if r<Wire_R
        B = 0 ##Update later.
    elseif r==Wire_R
        B=0
    else
        B = -1*μ₀*I / (2*π*r^2)
    end
    return B
end

File: Biot.jl-main/src_python/FieldCalc.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D  # 3Dプロット用
from BiotSav import biot_sav

def make_ellip(r1, r2, center=None, n_pts=100):
    """
    MakeEllip(r₁, r₂; Center=[0,0,0], NPts=100) の移植版
    楕円状の閉ループ上の点列 (N×3 ndarray) を返す
    """
    if center is None:
        center = np.array([0.0, 0.0, 0.0])
    else:
        center = np.asarray(center, dtype=float)
    theta = np.linspace(0, 2*np.pi, n_pts, endpoint=False)
    X = r1 * np.cos(theta) + center[0]
    Y = r2 * np.sin(theta) + center[1]
    Z = np.full(n_pts, center[2])
    return np.vstack([X, Y, Z]).T


def field_on_axis_circ(R, z, I=1.0):
    """
    FieldOnAxis_Circ(R, z; I=1) の移植版
    円形コイルの軸上磁場 Bz を返す
    """
    mu0 = 4 * np.pi * 1e-7
    return mu0 * 2*np.pi * R**2 * I / (4*np.pi * (z**2 + R**2)**1.5)


def make_ellip_test_points(r1, r2, center=None, n_pts=6, layers=3):
    """
    MakeEllipTestPoints(r₁, r₂; Center=[0,0,0], NPts=6, Layers=3) の移植版
    複数レイヤーの楕円状テスト点列を X, Y, Z の３つの 1D 配列で返す
    """
    if center is None:
        center = np.array([0.0, 0.0, 0.0])
    else:
        center = np.asarray(center, dtype=float)

    # レイヤー 1
    X = np.array([r1*1/(layers+1)*np.cos(i/n_pts*2*np.pi)+center[0] for i in range(n_pts)])
    Y = np.array([r2*1/(layers+1)*np.sin(i/n_pts*2*np.pi)+center[1] for i in range(n_pts)])
    Z = np.full(n_pts, center[2])

    # レイヤー 2..layers
    for ll in range(2, layers+1):
        x1 = np.array([r1*ll/(layers+1)*np.cos(i/n_pts*2*np.pi)+center[0] for i in range(n_pts)])
        y1 = np.array([r2*ll/(layers+1)*np.sin(i/n_pts*2*np.pi)+center[1] for i in range(n_pts)])
        z1 = np.full(n_pts, center[2])
        X = np.concatenate([X, x1])
        Y = np.concatenate([Y, y1])
        Z = np.concatenate([Z, z1])

    return X, Y, Z


def field_map_point_path(point_path, num_layers,
                         weight_radius=False, inv_weights=False,
                         plot_axes=None):
    """
    FieldMapPointPath(PointPath, NumLayers;
                      WeightRadius=false, InvWeights=false, PlotAxes=nothing) の移植版

    ・PointPath: (N×3) ndarray
    ・NumLayers: 内部テスト点レイヤ数
    ・weight_radius / inv_weights: 重み付けオプション
    ・plot_axes: matplotlib の Axes3D オブジェクトを渡せばそこにプロット
    """
    pts = np.asarray(point_path, dtype=float)
    x = pts[:,0]; y = pts[:,1]
    xTP = x.copy(); yTP = y.copy()

    # 内部レイヤー点を追加
    xmid, ymid = (x.min()+x.max())/2, (y.min()+y.max())/2
    for i in range(1, num_layers-1):
        factor = i/num_layers
        xTP = np.concatenate([xTP, (x - xmid)*factor + xmid])
        yTP = np.concatenate([yTP, (y - ymid)*factor + ymid])

    # テスト点リスト作成
    TP_list = np.vstack([xTP, yTP, np.zeros_like(xTP)]).T

    # 重み配列の生成
    if weight_radius:
        weights = x.copy()
    else:
        weights = np.ones_like(x)
    if inv_weights:
        weights = 1.0 / weights
    # 閉ループ用に先頭要素を末尾に追加
    weights = np.concatenate([weights, weights[:1]])

    # 各テスト点で BiotSav を呼び出し、Z成分の絶対値を収集
    Bmag = np.array([
        abs(biot_sav(pts.tolist(), tp.tolist(),
                     current=weights, min_threshold=1e-8)[2])
        for tp in TP_list
    ])

    # 0 でない点のみ抽出してプロット
    mask = (Bmag != 0)
    Xp, Yp, Bp = xTP[mask], yTP[mask], Bmag[mask]

    # 3Dサーフェスプロット
    if plot_axes is not None:
        ax = plot_axes
    else:
        fig = plt.figure()
        ax = fig.add_subplot(111, projection='3d')
    ax.plot_trisurf(Xp, Yp, Bp, cmap='jet')
    ax.set_xlabel('X'); ax.set_ylabel('Y'); ax.set_zlabel('|B_z|')


def wire_near_field_B(r, wire_R, I=1.0):
    """
    WireNearField_B(r, Wire_R; I=1) の移植版
    距離 r での近傍磁場 B を返す（スカラー or ndarray 対応）
    """
    mu0 = 4 * np.pi * 1e-7
    r_arr = np.asarray(r)
    # r < wire_R
    B = np.where(r_arr < wire_R,
                 mu0 * r_arr * I / (2*np.pi * wire_R**2),
    # r == wire_R
                 np.where(r_arr == wire_R,
                          mu0 * I / (2*np.pi * wire_R),
    # r > wire_R
                          mu0 * I / (2*np.pi * r_arr)))
    return B if B.shape else float(B)


def deriv_field(r, wire_R=0.001, I=1.0):
    """
    DerivField(r, Wire_R=0.001; I=1) の移植版
    磁場の半径微分 dB/dr を返す
    """
    mu0 = 4 * np.pi * 1e-7
    r_arr = np.asarray(r)
    # r > wire_R のみ微分計算、それ以外は 0
    Bp = np.where(r_arr > wire_R,
                  -mu0 * I / (2*np.pi * r_arr**2),
                  0.0)
    return Bp if Bp.shape else float(Bp)
